_posts/2023-01-19-blog-new-svg-engine.md

```diff
@@ -0,0 +1,201 @@
+
+By **July 2022**, after testing different approaches, a proper solution <a href="https://commits.webkit.org/252643@main">to add `viewBox` support was upstreamed</a>. The chosen solution makes use of another CSS concept that arises in the context of generated content: *"anonymous boxes"*. The idea is to wrap the direct descendants of `RenderSVGRoot` in an anonymous `RenderSVGViewportContainer` ("anonymous" = no associated DOM element) and apply the `viewBox` transformation as a regular CSS transformation on the anonymous renderer. With that approach, LBSE is left with just a *single, unified* `viewBox` implementation, without error-prone special cases in `RenderSVGRoot`, unlike the legacy SVG engine which has two disjoint implementations in `LegacyRenderSVGViewportContainer` and `LegacyRenderSVGRoot`.
+
+After the summer holidays, in **August 2022**, the next major milestone was reached: <a href="https://bugs.webkit.org/show_bug.cgi?id=242833">enabling compositing support for arbitrary SVG elements</a>, bringing **z-index** support, **hardware-accelerated compositing** and **3D transforms** to SVG. This time *all lessons* learned from the previous LBSE prototypes were taken into account, resulting in a *complete* compositing implementation, that works in various scenarios: different `transform-box` / `transform-origin` combinations, inline SVG enclosed by absolute/relative positioned CSS boxes and many more, all way more polished than in the "final" LBSE prototype.
+
+The aforementioned patch contained a fix for a <a href="https://webkit.org/b/27684">long-standing bug</a> (*"Composited elements appear pixelated when scaled up using transform"*), that made composited elements look blurry when scaling up with a CSS `transform` animation. The so-called *"backing scale factor"* of the associated `GraphicLayers` (see <a href="{{ '/blog/03-wpe-graphics-architecture.html' | url }}">here for details</a> about the role of `GraphicLayer` in the compositing system</a>) never changes during the animation. Therefore, the rendered image was scaled up instead of re-rendering the content at the right scale. SVG now enforces updates of that scale factor, to avoid blurry SVGs. The fix is *not* activated yet for HTML as that requires more thought - see the previously-linked bug report for details.
```

Comment:
> SVG now enforces updates of that scale factor, to avoid blurry SVGs.

Should that say “LBSE now enforces updates…”?

Reply:
Correct!

Reply:
Pushed a fix.